### PR TITLE
Sbt process testrunner

### LINKED ISSUE #152: Improve how much control we have over mutation runs
Right now there is no control over the process when running a mutation. This means when a mutation causes an infinite loop we can't do much about it. It'd be great to have more control over the process. 

I have 2 ideas about how we could do this:
- Creating a sub-process/agent that then calls the test tasks. Should the mutation run take too long we can kill the agent and start a new one. Downside is killing the agent doesn't kill the sub-process (I think). Also means (I think) the build tool would have to start for each agent.
- Starting tests by starting a new java process with the appropriate classpath, main class, etc. This means collecting the classpath and all that first, then calling `java -cp ...`. This gives a lot of control over the process (as much as possible, really). Downside is likely a lot more work to implement, and possibly test-runner specific work (Scalatest, JUnit, all have different entry points). This is also, for example, how IntelliJ runs tests. Looking at the intellij-scala plugin a lot of test-tool specific code happens there.

A combination of the two might also be possible, with agents starting new test runs.

There is some definite uncertainty for me in this area, as I am not very familiar with either of the two solutions.

### PR DESCRIPTION
**Pull request by hugo-vrijswijk**
### Closes #152 

Changes the sbt-plugin to run tests in a separate process. The plugin will start a new process with the new sbt-testrunner project and the application to test on the classpath. Also adds an api project to describe messages that are sent between the main and sub-processes.

The sbt-testrunner project will programmatically start tests using the [sbt test-interface API](https://github.com/sbt/test-interface). I made a small [POC to run tests programmatically](https://github.com/hugo-vrijswijk/test-init-poc) with the interface using either ScalaTest or MUnit. Both of those frameworks (and any testing frameworks that work in sbt) implement the test-interface.

To do:

- [x] Define projects
- [x] Start new java process from the main process
  - [x] Collect test context (test names to run, classpath, etc)
  - [x] Manage process in sbt plugin (starting, ending)
  - [x] Resolve correct scala-compatible version of child-process library (~~probably using [Coursier](https://get-coursier.io/docs/api)~~ adding the library to the `libraryDependencies` in sbt)
    - [x] Explore as many scalaVersions as we can for support. Currently Scala 2.11, 2.12, 2.13 and Dotty 0.27
- [x] Send messages to subprocess (probably using sockets?)
  - [x] ~~Experiment with Java RMI~~ Not needed. Serialization works well and keeps us in charge of the process
- [x] Start listening for messages in new test-process
- [x] Setup state for new process (tests to run, initialize framework, etc)
- [x] Handle messages (start tests, end process, etc)
- [x] Report results back from sub- to main-process
- [x] ~~Think of a better name for the sub-process package~~ Simply sbt-testrunner. If we expand it to maven the package can be renamed
- [x] Cleanup
- [x] Fix/improve/add tests
- [x] Fix/investigate issue with StackoverflowError after ~80 mutants (when running on core project)
- [x] Timeout detection (hardcoded duration)
- [x] Retry mechanism for starting new processes
- [x] Pass-through java arguments

Out of scope for this PR:

- Plugins that aren't sbt. 
  - This is probably possible for the maven plugin, but I want a working MVP first. And it would likely require additional setup for junit tests or other test frameworks as those don't implement the sbt test-interface
- Setups with multiple testing frameworks. Assumption is a single test-framework. Will probably work, but will be verified after this PR.
- Detecting static mutants (in objects, for example, the code will be run before any mutations are activated, so they can never be mutated). We should ignore these in the future
- Filtering tests on coverage
- Filtering tests on speed
- Starting multiple processes for performance
- Scala.js. JVM only for now

It also opens up the way towards features like #412, #312, #16 and #10 (wow that's a big list)

### CODE REVIEW

File: project/Dependencies.scala [hunk 1]
```diff
@@ -25,6 +26,7 @@ object Dependencies {
     val mockitoScala = "org.mockito" %% "mockito-scala-scalatest" % versions.mockitoScala % Test
   }
 
+  val testInterface = "org.scala-sbt" % "test-interface" % versions.testInterface
```
Comment (nicojs) [inline project/Dependencies.scala:29]: Curious. What happens if the project under test depends on a different version of this interface?

File: sbt-testrunner/src/main/scala/stryker4s/sbt/testrunner/SbtTestRunner.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    var status = acc
+    val eventHandler: EventHandler = event => {
+      status = combineStatus(status, event.status())
+    }
+    val newTasks = testTasks.flatMap(task => task.execute(eventHandler, Array.empty))
+    if (newTasks.nonEmpty) runMutation(newTasks, status)
+    else status
+  }
+
+  def combineStatus(current: Status, newStatus: Status) =
```
Comment (nicojs) [inline sbt-testrunner/src/main/scala/stryker4s/sbt/testrunner/SbtTestRunner.scala:41]: This def seems awfully verbose given what it's doing 🙈

File: sbt/src/main/scala/stryker4s/sbt/runner/ProcessHandler.scala [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package stryker4s.sbt.runner
+
+import scala.tools.nsc.io.Socket
+import stryker4s.model.MutantRunResult
+import java.nio.file.Path
+
+class ProcessHandler() {
+  def newProcess(classpath: Seq[String]) = {
+    val socket = new Socket(new java.net.Socket("127.0.0.1", 13337))
```
Comment (nicojs) [inline sbt/src/main/scala/stryker4s/sbt/runner/ProcessHandler.scala:9]: Old school. Inter process communication using sockets?

File: sbt/src/main/scala/stryker4s/sbt/runner/SbtMutantRunner.scala [hunk 4]
```diff
@@ -9,13 +9,11 @@ import sbt._
...
 import stryker4s.extension.FileExtensions._
-import stryker4s.extension.exception.InitialTestRunFailedException
 import stryker4s.model._
 import stryker4s.mutants.findmutants.SourceCollector
 import stryker4s.report.Reporter
 import stryker4s.run.MutantRunner
 import stryker4s.sbt.Stryker4sMain.autoImport.stryker
-import sbt.Tests.Output
 
 class SbtMutantRunner(state: State, sourceCollector: SourceCollector, reporter: Reporter)(implicit config: Config)
```
Comment (nicojs) [inline sbt/src/main/scala/stryker4s/sbt/runner/SbtMutantRunner.scala:18]: What is a `MutantRunner`? I'd imagine it is the `TestRunner` adapter for Stryker4s? Is it also used for the initial test run?

File: project/Dependencies.scala (hunk 1 above)
Comment (hugo-vrijswijk) [reply project/Dependencies.scala:29]: The last version of this package was published in 2013, so I would say that's a pretty stable API. Even so, I've now changed the dependency to 'Provided' (so assumed to be there). As long as there are no breaking changes used in the API it should be fine

File: sbt-testrunner/src/main/scala/stryker4s/sbt/testrunner/SbtTestRunner.scala (hunk 2 above)
Comment (hugo-vrijswijk) [reply sbt-testrunner/src/main/scala/stryker4s/sbt/testrunner/SbtTestRunner.scala:41]: Open for PR's 😊. Everything in this PR is very WIP at the moment

File: sbt/src/main/scala/stryker4s/sbt/runner/SbtMutantRunner.scala (hunk 4 above)
Comment (hugo-vrijswijk) [reply sbt/src/main/scala/stryker4s/sbt/runner/SbtMutantRunner.scala:18]: Yes and yes. Does this mean I've named it correctly? I'm not sure TBH. It's responsible for running tests with the correct mutation activated (there's one for sbt, maven, command-runner)

File: sbt/src/main/scala/stryker4s/sbt/runner/ProcessHandler.scala (hunk 3 above)
Comment (hugo-vrijswijk) [reply sbt/src/main/scala/stryker4s/sbt/runner/ProcessHandler.scala:9]: Yeah... If it works don't fix it? The issue is I need some way to talk to talk to a process that is:

- Two-way communication
- Not stdin. Anything can write to stdin using `println`
- Doesn't introduce new dependencies. Could cause conflicts with the system under test
- Able to serialize jvm objects through. JSON/protobuf/anything else introduces new dependencies, see point 2

So I ended up with sockets. It's very low-level, but I'm open to suggestions haha

**Comment by hugo-vrijswijk** [comment]
> I have more questions than anything else 😀

Exactly what I was hoping for

**Pull request merged**